hi-ml-histopathology/testhisto/testhisto/mocks/deepmil.py

```diff
@@ -0,0 +1,139 @@
+            is_finetune=False,
+            class_names=["ISUP 0", "ISUP 1", "ISUP 2", "ISUP 3", "ISUP 4", "ISUP 5"],
+            # Encoder parameters
+            encoder_type=ImageNetEncoder.__name__,
+            tile_size=28,
+            # Data Module parameters
+            batch_size=2,
+            max_bag_size=4,
+            max_bag_size_inf=4,
+            encoding_chunk_size=4,
+            cache_mode=CacheMode.NONE,
+            precache_location=CacheLocation.NONE,
+            # declared in DatasetParams:
+            local_datasets=[Path(tmp_path)],
+            # declared in TrainerParams:
+            max_epochs=2,
+            crossval_count=1,
+            # declared in OptimizerParams:
+            l_rate=5e-4,
+            weight_decay=1e-4,
+            adam_betas=(0.9, 0.99),
+        )
+        default_kwargs.update(kwargs)
+        super().__init__(**default_kwargs)
+
+        self.tmp_path = tmp_path
+        self.best_checkpoint_filename = "checkpoint_max_val_auroc"
+        self.best_checkpoint_filename_with_suffix = self.best_checkpoint_filename + ".ckpt"
+        self.checkpoint_folder_path = tmp_path / "outputs/checkpoints/"
+        best_checkpoint_callback = ModelCheckpoint(
+            dirpath=self.checkpoint_folder_path,
+            monitor="val/accuracy",
+            filename=self.best_checkpoint_filename,
+            auto_insert_metric_name=False,
+            mode="max",
+        )
+        self.callbacks = best_checkpoint_callback
+
+    @property
+    def cache_dir(self) -> Path:
+        return Path(self.tmp_path / f"innereye_cache1/{self.__class__.__name__}-{self.encoder_type}/")
+
+    def get_data_module(self) -> MockTilesDataModule:
+        image_key = MockTilesDataset.IMAGE_COLUMN
+        if self.is_finetune:
+            transform = LoadTilesBatchd(image_key, progress=True)
+            num_cpus = os.cpu_count()
+            assert num_cpus is not None  # for mypy
+            workers_per_gpu = num_cpus // torch.cuda.device_count()
+            dataloader_kwargs = dict(num_workers=workers_per_gpu, pin_memory=True)
+        else:
+            transform = Compose(
+                [
+                    LoadTilesBatchd(image_key, progress=True),
+                    EncodeTilesBatchd(image_key, self.encoder, chunk_size=self.encoding_chunk_size),
+                ]
+            )
+            dataloader_kwargs = dict(num_workers=0, pin_memory=False)
+
+        return MockTilesDataModule(
+            root_path=self.local_datasets[0],
+            max_bag_size=self.max_bag_size,
+            batch_size=self.batch_size,
+            max_bag_size_inf=self.max_bag_size_inf,
+            transform=transform,
+            cache_mode=self.cache_mode,
+            precache_location=self.precache_location,
+            cache_dir=self.cache_dir,
+            crossval_count=self.crossval_count,
+            crossval_index=self.crossval_index,
+            dataloader_kwargs=dataloader_kwargs,
+        )
```

Comment:
I think we're rewriting too much of the logic in this test. We risk the test to go out of sync with BaseMIL and at that point it would no longer be useful. My suggestion is 

- where possible to make use of `@mock.patch` t patch the methods/function we want to overwrite - here for example can we patch TilesDataModule with MockTilesDataModule?
- refactor bits of basemil, here for example you could have a separate class method that returns `transform, dataloader_kwargs` that incorporates the logic you're copying

Reply:
Good point, I'll look into that!